### DISCUSSION TITLE
Tracking when a commit or begin transaction is emitted to the db

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#7938] [Usage Questions] [answered]
I'd like to build a tool for myself that tracks whenever a begin transaction or commit is issued to the underlying database, at the most fundamental level of SQLAlchemy. This is so that I can better understand what Pythonic operations result in database activity. What would be the "right way" to do that? I'm assuming event listeners, but not sure on what component that would be most central to that goal.

Thanks in advance.

**Comment by zzzeek** [comment, marked as answer]
the most fundamental level is the connection, so use the commit/rollback event handlers: 

https://docs.sqlalchemy.org/en/14/core/events.html#sqlalchemy.events.ConnectionEvents.commit
https://docs.sqlalchemy.org/en/14/core/events.html#sqlalchemy.events.ConnectionEvents.rollback

the DBAPI has no "begin".  there is a "begin" handler as well but that does not actually do anything to the DBAPI connection, it just tracks when the Connection considers the transaction to have begun: https://docs.sqlalchemy.org/en/14/core/events.html#sqlalchemy.events.ConnectionEvents.begin    .  so to really see when the database server itself "begins", you'd need to coordinate between the "begin" event and the "before_cursor_execute" event: https://docs.sqlalchemy.org/en/14/core/events.html#sqlalchemy.events.ConnectionEvents.before_cursor_execute .
